File: test/Server/Data/PaginationTest.cs
namespace Belin.Base.Data;

/// <summary>
/// Tests the features of the <see cref="Pagination"/> class.
/// </summary>
[TestClass]
public sealed class PaginationTest {

	[TestMethod]
	public void CurrentPageIndex() {
		// It should always be greater than or equal to one.
		AreEqual(1, new Pagination { CurrentPageIndex = -1 }.CurrentPageIndex);
		AreEqual(1, new Pagination { CurrentPageIndex = 0 }.CurrentPageIndex);
		AreEqual(123, new Pagination { CurrentPageIndex = 123 }.CurrentPageIndex);
	}

	[TestMethod]
	public void HasNextPage() {
		// It should return `false` if there is no next page.
		IsFalse(new Pagination().HasNextPage);

		// It should return `true` if a next page exists.
		IsTrue(new Pagination { CurrentPageIndex = 2, TotalItemCount = 123 }.HasNextPage);
	}

	[TestMethod]
	public void HasPreviousPage() {
		// It should return `false` if there is no previous page.
		IsFalse(new Pagination().HasPreviousPage);

		// It should return `true` if a previous page exists.
		IsTrue(new Pagination { CurrentPageIndex = 5 }.HasPreviousPage);
	}

	[TestMethod]
	public void ItemsPerPage() {
		// It should always be between 1 and 1000.
		AreEqual(1, new Pagination { ItemsPerPage = -1 }.ItemsPerPage);
		AreEqual(1, new Pagination { ItemsPerPage = 0 }.ItemsPerPage);
		AreEqual(1000, new Pagination { ItemsPerPage = 9999 }.ItemsPerPage);
	}

	[TestMethod]
	public void LastPageIndex() {
		// It should return the total count divided by the page size rounded up.
		AreEqual(0, new Pagination { TotalItemCount = 0 }.LastPageIndex);
		AreEqual(123, new Pagination { ItemsPerPage = 1, TotalItemCount = 123 }.LastPageIndex);
		AreEqual(3, new Pagination { ItemsPerPage = 10, TotalItemCount = 25 }.LastPageIndex);
	}

	[TestMethod]
	public void Offset() {
		// It should return the page size multiplied by the page index minus one.
		AreEqual(0, new Pagination { CurrentPageIndex = 1 }.Offset);
		AreEqual(100, new Pagination { CurrentPageIndex = 5, ItemsPerPage = 25 }.Offset);
		AreEqual(610, new Pagination { CurrentPageIndex = 123, ItemsPerPage = 5 }.Offset);
	}

	[TestMethod]
	public void TotalItemCount() {
		// It should always be greater than or equal to zero.
		AreEqual(0, new Pagination { TotalItemCount = -1 }.TotalItemCount);
		AreEqual(0, new Pagination { TotalItemCount = 0 }.TotalItemCount);
		AreEqual(123, new Pagination { TotalItemCount = 123 }.TotalItemCount);
	}
}

File: test/Client/Data/PaginationTest.js
import {Pagination} from "@cedx/base/Data/Pagination.js";
import {assert} from "chai";

/**
 * Tests the features of the {@link Pagination} class.
 */
describe("Pagination", () => {
	describe("lastPageIndex", () => {
		it("should return the total count divided by the page size rounded up", () => {
			assert.equal(new Pagination({totalItemCount: 0}).lastPageIndex, 0);
			assert.equal(new Pagination({itemsPerPage: 1, totalItemCount: 123}).lastPageIndex, 123);
			assert.equal(new Pagination({itemsPerPage: 10, totalItemCount: 25}).lastPageIndex, 3);
		});
	});

	describe("offset", () => {
		it("should return the page size multiplied by the page index minus one", () => {
			assert.equal(new Pagination({currentPageIndex: 1}).offset, 0);
			assert.equal(new Pagination({currentPageIndex: 5, itemsPerPage: 25}).offset, 100);
			assert.equal(new Pagination({currentPageIndex: 123, itemsPerPage: 5}).offset, 610);
		});
	});

	describe("searchParams", () => {
		it("should include a `page` parameter", () => {
			assert.equal(new Pagination().searchParams.get("page"), "1");
			assert.equal(new Pagination({currentPageIndex: -5}).searchParams.get("page"), "1");
			assert.equal(new Pagination({currentPageIndex: 123}).searchParams.get("page"), "123");
		});

		it("should include a `perPage` parameter", () => {
			assert.equal(new Pagination().searchParams.get("perPage"), "25");
			assert.equal(new Pagination({itemsPerPage: 66}).searchParams.get("perPage"), "66");
			assert.equal(new Pagination({itemsPerPage: 123456}).searchParams.get("perPage"), "1000");
		});
	});

	describe("fromResponse()", () => {
		it("should create an instance initialized from the response headers", () => {
			const pagination = Pagination.fromResponse(new Response(null, {headers: {
				"X-Pagination-Current-Page": "123",
				"X-Pagination-Per-Page": "33",
				"X-Pagination-Total-Count": "666"
			}}));

			assert.equal(pagination.currentPageIndex, 123);
			assert.equal(pagination.itemsPerPage, 33);
			assert.equal(pagination.totalItemCount, 666);
		});
	});
});
